### PR TITLE
add main_module attribute (adds new mappings)

### PR DESCRIPTION
**Pull request by andreeap**
This is a pull request for #149 , but it's still in progress. I also need to add a test.

*requires new mappings*

### CODE REVIEW

File: lib/MetaCPAN/Script/Release.pm [hunk 1]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
 
+sub find_main_module {
+    my $self = shift;
+    my ( $mod, $distribution ) = @_;
+
+    my @modules = @{$mod};
+
+    my $min_level          = 100;
+    my @distribution_parts = split /-/, $distribution;
+    my $distribution_name  = join '', @distribution_parts;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: If you want to compare the distribution name to the module name, you could probably get a comparison variable without the split + join: 

```
my $dist2module = $distribution;
$dist2module =~ s{-}{::}g;
```

Now `HTML-Restrict` becomes `HTML::Restrict`.  If that matches a module, you've got your main_module.

File: lib/MetaCPAN/Script/Release.pm [hunk 2]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
+
+    if ( scalar @modules eq 1 ) {
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: `if ( scalar @modules == 1 ) {`

You want a numeric comparison (not string).  Even if it actually works, using the `==` is less confusing.

File: lib/MetaCPAN/Script/Release.pm [hunk 3]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
+
+        # there is only one module and it will become the main_module
+        my $main_module = shift @modules;
+        $self->main_module( $main_module->module );
+        return;
+    }
+    else {
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: Since you're returning in the `if` above, the `else` is not necessary.  If you get rid of it, you can save yourself one level of indentation.

File: lib/MetaCPAN/Script/Release.pm [hunk 4]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
+        foreach my $file ( sort @modules ) {
+            my @module_parts = split /::/, $file->module->[0]->{name};
+            my $module_name = join '', @module_parts;
+
+            # the module has the exact name as the ditribution
+            if ( $module_name eq $distribution_name ) {
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: Using the code I suggested above: `if ( $file eq $dist2module ) {`

File: lib/MetaCPAN/Script/Release.pm [hunk 5]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
+
+        # there is only one module and it will become the main_module
+        my $main_module = shift @modules;
+        $self->main_module( $main_module->module );
+        return;
+    }
+    else {
+        foreach my $file ( sort @modules ) {
+            my @module_parts = split /::/, $file->module->[0]->{name};
+            my $module_name = join '', @module_parts;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: You can delete these two lines.

File: lib/MetaCPAN/Script/Release.pm [hunk 6]
```diff
@@ -490,6 +498,73 @@ sub import_tarball {
...
+
+            # the module contains the ditribution name
+            if ( $module_name =~ /$distribution_name/ ) {
+                $self->main_module( $file->module );
+                return;
+            }
+
+            # the distribution has modules on different levels
+            # the main_module is the one with minimum level
+            if ( $file->level lt $min_level ) {
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: I'm not going to look at this logic too closely right now.  I think after you add some tests it will become clearer if this is the right approach.

File: lib/MetaCPAN/Script/Release.pm [hunk 7]
```diff
@@ -490,6 +498,69 @@ sub import_tarball {
...
+
+    my $min_level   = 100;
+    my $dist2module = $distribution;
+    $dist2module =~ s{-}{::}g;
+
+    if ( scalar @modules == 1 ) {
+
+        # there is only one module and it will become the main_module
+        my $main_module = shift @modules;
+        $self->main_module( $main_module->module );
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: `$self->main_module( $modules[0] );`

**Comment by andreeap** [comment]
Yes, I remember you said to write it like that, but I left it this way because otherwise it won't get into the is() statement and it will give me 'Can't call method "main_module" on unblessed reference at t/script/release/find_main_module.t line 27.' and I wanted to point out that I can't get the main_module value out of the release script.

**Comment by rwstauner** [comment]
Why is `main_module` an attribute on the Script?  The script can process more than one release, and an attribute for the script is mainly intended to be passed in on the command line.

Can't we just assign the release's main_module attribute to the return value from a method call rather than caching it on the Script object first?
`find_main_module` could return the relevant value (instead of calling the attribute setter and then returning nothing)
and then:
`$release->main_module( $self->find_main_module( \@modules, $release->distribution ) );`

**Comment by andreeap** [comment]
I modified the logic for choosing the main_module and also set it only for  $release without caching the value in the script. The main_module is set correctly for the example in the test, but I hardcoded the value because I am not sure how to access it there(I just take a tarball and pass it to the release script without doing something else). Should I rethink the test?

**Comment by rwstauner** [comment]
Yes, let's rethink the test.
I see no reason that we need to explicitly call the Script... all we really want to test is that the release is saved with the expected main_module attribute.

We already have lots of tests in `t/release` that test our fake distributions.
Rather than writing up a custom test for this I'd suggest just adding "main_module" checks to our existing release tests.  Then we can add more fake releases if we need to test specific/edge cases rather than seeking out more "real" dists to have to download.
That will be more consistent with our existing test and much simpler (we won't need the `Test::RequiresInternet` and the boilerplate you had to figure out to run the Script).

We should be able to just remove the new test file you added (and the new dependencies in cpanfile\*)
and instead add a line like:
`main_module => 'BadPod',`
to files like `t/release/badpod.t`.

That being said, I tried that, and discovered some issues with the code that we should discuss.

My first question: Do we want "main_module" to be a module name like "Foo::Bar", or the "module" hashref like `{ name => "Foo::Bar", associated_pod => "" ... }`.
The latter is what it's being set to now (based on `MetaCPAN::Types::Internal::Module`), but the former is what it's testing for.
Originally I thought we would use just the name, but if anyone feels differently I'm open to discussion.

**Comment by andreeap** [comment]
In order to test for the main_module, can I change this to only setting the name of the module or keep it the way it is now?

**Comment by andreeap** [comment]
I set only the name of a module as the main_module, and when the releases from fakecpan are indexed, main_module is set, but when it gets to t/release/<release_name.t> this attribute is left undefined. What am I missing?

**Comment by oalders** [comment]
> Originally I thought we would use just the name, but if anyone feels differently I'm open to discussion

I think we do just want to use the name rather than a HashRef here.  Good catch!

**Comment by rwstauner** [comment]
@andreeap unfortunately that method (that whole file) is a bit confusing.  It could use some cleanup...
Primarily I see no obvious reason why there are multiple calls to `$release->put` (there are 3 within 20 lines of code), all inside if statements. That could surely be improved.
That's not something to fix in this pull request, however it's related to the issue.

You are calling `set_main_module` _after_ the last `put` so the document never makes it to the index.
In order to persist the change, I would suggest moving the call to `set_main_module` before the `$tmpdir->rmtree` line (before passing to the "latest" script), and then calling `$release->put` to make sure it is saved.

We can deal with cleaning up the confusion later (i think there are already outstanding branches about cleaning this up).

So, to clarify, I would suggest a change like this:

```diff
diff --git a/lib/MetaCPAN/Script/Release.pm b/lib/MetaCPAN/Script/Release.pm
index 02ef435..4f95d8c 100644
--- a/lib/MetaCPAN/Script/Release.pm
+++ b/lib/MetaCPAN/Script/Release.pm
@@ -481,16 +481,18 @@ sub import_tarball {
         $release->authorized(0);
         $release->put;
     }
 
+    $self->set_main_module( \@modules, $release );
+    $release->put;
+
     $tmpdir->rmtree;
 
     if ( $self->latest ) {
         local @ARGV = ( qw(latest --distribution), $release->distribution );
         MetaCPAN::Script::Runner->run;
     }
 
-    $self->set_main_module( \@modules, $release );
 }
 
 sub set_main_module {
     my $self = shift;

```

**Comment by andreeap** [comment]
I rebased this pull request and added the field to some of the tests and it is ready to be reviewed.

**Comment by oalders** [comment]
Looks good to me.  @rwstauner, any objections?  If not, I'll merge it tomorrow (Friday).

**Comment by rwstauner** [comment]
@andreeap, thanks for the effort you put into updating all those tests.
This looks great!
